### PR TITLE
fleet2.0 add fp16 grad compression

### PR DESCRIPTION
**Pull request by wangxicoding**
### PR types

New features

### PR changes

APIs

### Describe

Old PR https://github.com/PaddlePaddle/Paddle/pull/22434 , rewrite in this PR.
Add fp16 grad compress to fleet meta optimizer. The strategy name is `fp16_allreduce`
Test on ResNet50 model，run with `fleetrun --gpus 0,1,2,3 fp16_compress_demo.py`
``` python
import paddle
paddle.enable_static()

import fleetx as X
import paddle.fluid as fluid
import paddle.distributed.fleet as fleet


fleet.init(is_collective=True)
configs = X.parse_train_configs()
model = X.applications.Resnet50("NCHW")

imagenet_downloader = X.utils.ImageNetDownloader()
local_path = imagenet_downloader.download_from_hdfs('imagenet.yaml',local_path='./ImageNet')
loader = model.load_imagenet_from_file("{}/train.txt".format(local_path), batch_size=32)

dist_strategy = fleet.DistributedStrategy()
dist_strategy.fp16_allreduce = True

optimizer = fluid.optimizer.Momentum(learning_rate=configs.lr, momentum=configs.momentum)
optimizer = fleet.distributed_optimizer(optimizer, strategy=dist_strategy)
optimizer.minimize(model.loss)

trainer = X.MultiGPUTrainer()
trainer.fit(model, loader, epoch=10, use_dali=True)

```

### CODE REVIEW

File: python/paddle/distributed/fleet/base/distributed_strategy.py [hunk 1]
```diff
@@ -845,6 +845,30 @@ def dgc_configs(self, configs):
...
+            strategy = fleet.DistributedStrategy()
+            strategy.fp16_compress = True # by default this is false
+
+        """
+        return self.strategy.fp16_compress
+
+    @fp16_compress.setter
+    @is_strict_auto
+    def fp16_compress(self, flag):
+        if isinstance(flag, bool):
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/base/distributed_strategy.py]: assert

File: python/paddle/fluid/tests/unittests/test_fleet_distributed_strategy.py [hunk 2]
```diff
@@ -102,6 +102,15 @@ def test_dgc(self):
         strategy.dgc = "True"
         self.assertEqual(strategy.dgc, False)
 
+    def test_fp16_compress(self):
+        strategy = paddle.distributed.fleet.DistributedStrategy()
+        strategy.fp16_compress = True
+        self.assertEqual(strategy.fp16_compress, True)
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_fleet_distributed_strategy.py]: 这是测试啥？

File: python/paddle/distributed/fleet/base/distributed_strategy.py (hunk 1 above)
Comment (wangxicoding) [reply python/paddle/distributed/fleet/base/distributed_strategy.py]: Done. raise TypeError

File: python/paddle/fluid/tests/unittests/test_fleet_distributed_strategy.py (hunk 2 above)
Comment (wangxicoding) [reply python/paddle/fluid/tests/unittests/test_fleet_distributed_strategy.py]: 测试distributed_strategy的各个属性是否生效

File: python/paddle/distributed/fleet/meta_optimizers/fp16_compress_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                backward = op_maker.OpRole.Backward
+                cast_op._set_attr(op_maker.kOpRoleAttrName(), backward)
+                cast_op._set_attr(op_maker.kOpRoleVarAttrName(),
+                                  [param.name, new_grad.name])
+                new_grad.op = cast_op
+
+            new_param_and_grads.append((param, new_grad, True))
+
+        ret_param_and_grads = []
+        # cast grad from fp16->fp32 after allreduce.
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_optimizers/fp16_compress_optimizer.py]: add a comment why split these to two segments.

Comment (wangxicoding) [reply python/paddle/distributed/fleet/meta_optimizers/fp16_compress_optimizer.py]: Done.

File: paddle/fluid/framework/distributed_strategy.proto [hunk 4]
```diff
@@ -127,6 +127,7 @@ message DistributedStrategy {
   optional int32 conv_workspace_size_limit = 22 [ default = 4000 ];
   optional bool cudnn_batchnorm_spatial_persistent = 23 [ default = true ];
   optional bool adaptive_localsgd = 24 [ default = false ];
+  optional bool fp16_compress = 25 [ default = false ];
```
Comment (guru4elephant) [inline paddle/fluid/framework/distributed_strategy.proto]: fp16_compress看起来信息冗余了，fp16本身就是压缩了，而且这个策略如果是通信策略，是不是叫fp16_allreduce之类的更好？

**Review by guru4elephant — COMMENTED** [review]
good PR, can you list some models that can benefit from your strategy?

File: paddle/fluid/framework/distributed_strategy.proto (hunk 4 above)
Comment (wangxicoding) [reply paddle/fluid/framework/distributed_strategy.proto]: Done. 已改名

**Comment by wangxicoding** [comment]
> good PR, can you list some models that can benefit from your strategy?

It has been tested before. Under the P4 card 10Gbps network, the accuracy can be aligned and the speed can be improved.
The detailed data will be retested later, including resnet, bert model, 100Gbps and 10Gbps bandwidth conditions.

**Review by gongweibao — APPROVED** [review]
LGTM
